nav2_util/include/nav2_util/geometry_utils.hpp

```diff
@@ -179,6 +219,120 @@ inline double calculate_path_length(const nav_msgs::msg::Path & path, size_t sta
   return path_length;
 }
 
+inline std::pair<geometry_msgs::msg::PoseStamped, geometry_msgs::msg::PoseStamped>
+find_closest_path_segment(
+  const nav_msgs::msg::Path & path,
+  const geometry_msgs::msg::PoseStamped & pose)
+{
+  if (path.poses.empty()) {
+    throw std::invalid_argument("No poses on path to project onto");
+  }
+
+  if (path.poses.size() == 1u) {
+    return std::make_pair(*path.poses.begin(), *path.poses.begin());
+  }
+
+  std::vector<double> distances(path.poses.size());
+  std::transform(
+    path.poses.begin(),
+    path.poses.end(),
+    distances.begin(),
+    [&pose](const auto & path_pose) {
+      return squared_euclidean_distance(path_pose, pose);
+    }
+  );
+  // We want to find the first local minima,
+  // so we don't loop around and get something from later
+  // if we start out by increasing distance away,
+  // then we will accept the beginning
+  auto distance_it = distances.begin();
+  for (; distance_it != std::prev(distances.end()); ++distance_it) {
```

Comment:
I don't think this necessarily works. Imagine you have a path going laterally then does a 90 degree turn up. The robot is somewhere in the upward portion of the path.

If the path going laterally has even a slight disturbance in it, then the differences in distances could be increasing by a slight amount to trigger this. 

Further, in the `std::transform` you're already owning an entire path-wise iteration. Why not just own it then and do the logic in that loop instead of using the transform function? This is actually now doing nearly double the compute, in the worst case.

Reply:
You may be right that this algorithm does not yield the closest segment to the path in general. However, for the context in which it is used, so far it has worked quite well, since RPP already continuously prunes up to the waypoint closest to the robot. So this is just a search after the fact, given the assumption that the robot is within a few waypoints of the beginning of the path. I think your counter-example is correct, but it would also be extremely unlikely given how RPP is used. However, if cross-track error is to be generalized, then I think this would need to be fixed.

Yes, these calculations could be consolidated into one for loop. The std::transform isn't needed.

Reply:
Thanks for the ideas. I will take a closer look at this, and when I do I will try to come up with some unit tests that test this.
- [ ] Make projected pose algorithm more robust
- [ ] Add unit tests for different paths to project onto.